### PR TITLE
(closes #1370) Adds 'collapse' and 'ignore_dependencies_for' options to parallel_loop_trans

### LINKED ISSUE #1370: OpenMP loop functions don't support collapse
The OMP collapse clause is not implemented for all directives and it can cause exceptions in transformations apply.

I believe the relevant transformations are OMPLoopTrans and OMPTaskloopTrans.

### PR DESCRIPTION
**Pull request by sergisiso**
Note that I changed the definition of the collapse option when it is an integer to mean collapse "upto x loops". Before it was collapse exactly 'x' loops or fail. But I found it hard to use for programmatical changing complex code as loops comes in all shapes. It could also be "atleast x loops" or fail.

Regardless, in practice in NEMO we always use the boolean option which means do as much as possible.

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter This is ready for a first review.

Note that the NEMO openmp for CPU is slower, Glados had several users at the time that may explain the behaviour, but I also removed some "forced" loop parallelisations that may impact the performance (I will repeat the integration test when its idle and if the slowdown is still that big, I will take a more nuanced approach).

Also you may want to check if the changes to ACCParallelTrans have any impact to the LFRic GPU scripts that are not in the repository.

File: examples/nemo/scripts/utils.py [hunk 1]
```diff
@@ -245,33 +245,33 @@ def insert_explicit_loop_parallelism(
...
-        opts = {}
+        opts = {"collapse": collapse, "verbose": True}
 
         routine_name = loop.ancestor(Routine).name
 
         if ('dyn_spg' in routine_name and len(loop.walk(Loop)) > 2):
-            print("Loop not parallelised because its in 'dyn_spg' and "
-                  "its not the inner loop")
+            loop.append_preceding_comment(
+                "PSyclone: Loop not parallelised because its in 'dyn_spg' and "
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:271]: s/its/it is/

File: examples/nemo/scripts/utils.py [hunk 2]
```diff
@@ -245,33 +245,33 @@ def insert_explicit_loop_parallelism(
...
+                "its not the inner loop")
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:272]: s/its not/is not/

File: examples/nemo/scripts/utils.py [hunk 3]
```diff
@@ -287,93 +287,43 @@ def insert_explicit_loop_parallelism(
...
-                return True
-            return False
-
         # If we see one such ice linearised loop, we assume
         # calls/codeblocks are not a problem (they are not)
-        if not any(ref.symbol.name in ('npti',)
-                   for ref in loop.stop_expr.walk(Reference)):
-            if skip_for_correctness(loop):
-                continue
+        # if any(ref.symbol.name in ('npti',)
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:305]: Remove altogether or extend comment to explain why it is commented out?

File: examples/nemo/scripts/utils.py [hunk 4]
```diff
@@ -287,93 +287,43 @@ def insert_explicit_loop_parallelism(
...
+        #            for ref in loop.stop_expr.walk(Reference)):
+        #     opts = {"force": True}
 
         # pnd_lev requires manual privatisation of ztmp
```
Comment (arporter) [inline examples/nemo/scripts/utils.py:309]: This is not new but presumably just ignoring the deps on 'ztmp' doesn't fix the need for manual privatisation?

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 5]
```diff
@@ -83,24 +85,30 @@ def validate(self, node, options=None):
...
-                        number of nested loops to collapse.
+        :param options: a dictionary with options for transformations.
         :type options: Optional[Dict[str, Any]]
-        :param int options["collapse"]: number of nested loops to collapse
-                                        or None.
+        :param bool|int options["collapse"]: if it's a bool and is False
+            (default), it won't collapse. If it's a bool and is True, it will
+            collapse as much as possible. If it's an integer, it will attempt
+            to collapse until the specified number of loops (if they exist and
+            are safe to collapse them). The options 'ignore_dependencies_for'
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:94]: s/them//

Comment (arporter) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:94]: Also, does this work outermost->inwards or innermost->outwards?

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 6]
```diff
@@ -83,24 +85,30 @@ def validate(self, node, options=None):
...
+            and 'force' also affect the collapse applicabilty analysis.
         :param bool options["force"]: whether to force parallelisation of the
-                target loop (i.e. ignore any dependence analysis).
+            target loop (i.e. ignore any dependence analysis).
+        :param list[str] options["ignore_dependencies_for"]: whether to ignore
+            some symbol names from the dependence analysis checks.
         :param bool options["sequential"]: whether this is a sequential loop.
+        :param bool options["verbose"]: whether report the reasons the validate
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:101]: "whether to report"

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 7]
```diff
@@ -83,24 +85,30 @@ def validate(self, node, options=None):
...
-                :py:class:`psyclone.psyir.nodes.Loop` loop iterates over \
-                colours.
-        :raises TransformationError: if 'collapse' is supplied with an \
-                invalid number of loops.
-        :raises TransformationError: if there is a data dependency that \
-                prevents the parallelisation of the loop unless \
-                `options["force"]` is True.
+        :raises TransformationError: if the given loop iterates over a
+                colours (LFRic domain) iteration space.
+        :raises TransformationError: if 'collapse' is not an int or a bool.
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:106]: As you mention in #2668, I think this should be a TypeError, not a TransformationError.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 8]
```diff
@@ -83,24 +85,30 @@ def validate(self, node, options=None):
...
+        :raises TransformationError: if 'ignore_dependencies_for' is not a
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:107]: TypeError again.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 9]
```diff
@@ -123,45 +133,45 @@ def validate(self, node, options=None):
         # If 'collapse' is specified, check that it is an int and that the
         # loop nest has at least that number of loops in it
         if collapse:
-            if not isinstance(collapse, int):
+            if not isinstance(collapse, (int, bool)):
                 raise TransformationError(
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:137]: TypeError here I think.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 10]
```diff
@@ -123,45 +133,45 @@ def validate(self, node, options=None):
...
 
         # Check that there are no loop-carried dependencies
         if sequential or ignore_dep_analysis:
             return
 
+        if ignore_dependencies_for:
+            if (not isinstance(ignore_dependencies_for, Iterable) or
+                    isinstance(ignore_dependencies_for, str) or not all(
+                    isinstance(v, str) for v in ignore_dependencies_for)):
+                raise TransformationError(
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:149]: TypeError again please.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 11]
```diff
@@ -123,45 +133,45 @@ def validate(self, node, options=None):
...
+                    f"The 'ignore_dependencies_for' option must be an Iterable"
+                    f" object containing containing str representing the "
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:151]: Just one "containing" :-)

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 12]
```diff
@@ -123,45 +133,45 @@ def validate(self, node, options=None):
...
             for message in dep_tools.get_all_messages():
                 if message.code == DTCode.WARN_SCALAR_WRITTEN_ONCE:
                     continue
                 all_msg_str = [str(message) for message in
                                dep_tools.get_all_messages()]
-                messages = "\n".join(all_msg_str)
-                raise TransformationError(
-                    f"Dependency analysis failed with the following "
-                    f"messages:\n{messages}")
+                messages = ("Loop can not be parallelised "
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:169]: s/can not/cannot/

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 13]
```diff
@@ -181,30 +191,101 @@ def apply(self, node, options=None):
...
+            collapse as much as possible. If it's an integer, it will attempt
+            to collapse until the specified number of loops (if they exist and
+            are safe to collapse them). The options 'ignore_dependencies_for'
+            and 'force' also affect the collapse applicabilty analysis.
+        :param bool options["force"]: whether to force parallelisation of the
+            target loop (i.e. ignore any dependence analysis).
+        :param list[str] options["ignore_dependencies_for"]: whether to ignore
+            some symbol names from the dependence analysis checks.
+        :param bool options["sequential"]: whether this is a sequential loop.
+        :param bool options["verbose"]: whether report the reasons the validate
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:210]: "whether to report"

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 14]
```diff
@@ -181,30 +191,101 @@ def apply(self, node, options=None):
...
+            next_loop = node
+            previous_iteration_variables = []
+            while isinstance(next_loop, Loop):
+                previous_iteration_variables.append(next_loop.variable)
+                num_collapsable_loops += 1
+                if not isinstance(collapse, bool):
+                    if num_collapsable_loops >= collapse:
+                        break
+
+                # If it has more than one children, the next loop will not be
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:243]: s/children/child/

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 15]
```diff
@@ -181,30 +191,101 @@ def apply(self, node, options=None):
...
+                # perfectly nested, so stop searching. If there is no child,
+                # we have an empty loop (which would cause a crash when
+                # accessing the child next)
+                if len(next_loop.loop_body.children) != 1:
+                    break
+
+                next_loop = next_loop.loop_body.children[0]
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:250]: Could just do `...loop_body[0]`.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 16]
```diff
@@ -181,30 +191,101 @@ def apply(self, node, options=None):
...
+                if not isinstance(next_loop, Loop):
+                    break
+
+                # If it is a loop dependent on a previous iteration variable
+                # (e.g. a triangular iteration space), it can not be collapsed
+                dependent_of_previous_variable = False
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:256]: "dependent_on_previous_variable"

File: src/psyclone/psyir/transformations/parallel_loop_trans.py [hunk 17]
```diff
@@ -181,30 +191,101 @@ def apply(self, node, options=None):
...
+                for bound in (next_loop.start_expr, next_loop.stop_expr,
+                              next_loop.step_expr):
+                    for ref in bound.walk(Reference):
+                        if ref.symbol in previous_iteration_variables:
+                            dependent_of_previous_variable = ref.symbol
+                            break
+                if dependent_of_previous_variable:
+                    if verbose:
+                        next_loop.preceding_comment = (
+                            f"Loop can not be collapsed because one of the "
```
Comment (arporter) [inline src/psyclone/psyir/transformations/parallel_loop_trans.py:266]: "cannot"

File: src/psyclone/psyir/nodes/omp_directives.py [hunk 18]
```diff
@@ -2546,7 +2546,7 @@ def validate_global_constraints(self):
                 f"OMPParallelDirective, but '{self}' is not.")
 
         # If there is a collapse clause, there must be as many immediately
-        # nested loops as the collapse value
+        # nested loops as the collapse value (minus itself)
```
Comment (arporter) [inline src/psyclone/psyir/nodes/omp_directives.py:2549]: I think the comment was correct as it was?

File: src/psyclone/psyir/tools/dependency_tools.py [hunk 19]
```diff
@@ -779,12 +779,15 @@ def can_loop_be_parallelised(self, loop,
...
-            var_info = var_accesses[signature]
-            symbol_table = loop.scope.symbol_table
-            symbol = symbol_table.lookup(var_name)
+            # Access the symbol by inspecting the first access reference
+            try:
+                symbol = var_info.all_accesses[0].node.symbol
+            except AttributeError:
+                # If its a node without a symbol, look it up
+                var_name = signature.var_name
+                symbol_table = loop.scope.symbol_table
```
Comment (arporter) [inline src/psyclone/psyir/tools/dependency_tools.py:788]: This line could be hoisted to L769.

File: src/psyclone/psyir/transformations/omp_target_trans.py [hunk 20]
```diff
@@ -86,12 +87,34 @@ class OMPTargetTrans(RegionTrans):
     '''
     excluded_node_types = (CodeBlock, )
 
+    def validate(self, node, options=None):
+        # pylint: disable=signature-differs
+        '''
+        Check that we can safely enclose the supplied node or list of nodes
+        within OpenACC kernels ... end kernels directives.
```
Comment (arporter) [inline src/psyclone/psyir/transformations/omp_target_trans.py:94]: this isn't a 'kernels' transformation! :-)

File: src/psyclone/psyir/transformations/omp_target_trans.py [hunk 21]
```diff
@@ -86,12 +87,34 @@ class OMPTargetTrans(RegionTrans):
...
+        :type options: Optional[Dict[str, Any]]
+        '''
+        node_list = self.get_node_list(node)
+        super().validate(node, options)
+        for node in node_list:
+            for call in node.walk(Call):
+                if not call.is_available_on_device():
+                    raise TransformationError(
+                        f"'{call.routine.name}' is not available on the "
+                        f"accelerator device, and therefore it can not "
```
Comment (arporter) [inline src/psyclone/psyir/transformations/omp_target_trans.py:109]: "cannot"

File: src/psyclone/psyir/transformations/omp_target_trans.py [hunk 22]
```diff
@@ -86,12 +87,34 @@ class OMPTargetTrans(RegionTrans):
...
+                        f"be enclosed in a OMP target region.")
```
Comment (arporter) [inline src/psyclone/psyir/transformations/omp_target_trans.py:110]: "be called from within an OMP target region."

File: src/psyclone/psyir/transformations/omp_target_trans.py [hunk 23]
```diff
@@ -86,12 +87,34 @@ class OMPTargetTrans(RegionTrans):
...
+        '''
+        Check that we can safely enclose the supplied node or list of nodes
+        within OpenACC kernels ... end kernels directives.
+
+        :param node: the PSyIR node or nodes to enclose in the OpenMP
+                      target region.
+        :type node: List[:py:class:`psyclone.psyir.nodes.Node`]
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, Any]]
+        '''
```
Comment (arporter) [inline src/psyclone/psyir/transformations/omp_target_trans.py:101]: Please document the :raises:

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 24]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+    psyir = fortran_reader.psyir_from_source(CODE)
+    loop = psyir.walk(Loop)[0]
+    trans = ParaTrans()
+    with pytest.raises(TransformationError) as err:
+        trans.validate(loop, options={"verbose": True})
+    assert ("Loop can not be parallelised because the dependency analysis "
+            "reported:\nWarning: Variable 'sum' is read first, which indicates"
+            " a reduction. Variable: 'sum'.") in str(err.value)
+    # With the verbose option, the dependency issue will be log as a comment
+    print(fortran_writer(psyir))
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:107]: Please rm this line.

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 25]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+    'ignore_dependencies_for' options.
+    '''
+    psyir = fortran_reader.psyir_from_source('''
+        subroutine my_sub()
+          integer :: i, j, k
+          real :: var(10,10,10) = 1
+          integer, dimension(10) :: map
+
+          do i = 1, 10  ! This loop is iteration independent
+            do j = 1, 10  ! This loop has a loop-carried dependency in var1
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:167]: s/var1/var

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 26]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+              do k = 1, j  ! This loop bounds depend on previous indices
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:168]: "...loop bound depends on..."

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 27]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+    trans.validate(loop)
+
+    # By default it stops at collapse 1, because loop 2 has a dependency on,
+    # var. With verbose it adds the stopping reason as a comment.
+    test_loop = psyir.copy().walk(Loop, stop_type=Loop)[0]
+    trans.apply(test_loop, {"collapse": True, "verbose": True})
+    assert '''\
+!$omp parallel do collapse(1) default(shared), private(i,j,k)
+do i = 1, 10, 1
+  ! Error: The write access to 'var(i,j,k)' and the read access to 'var(i,\
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:187]: Very nice :-)

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 28]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+              end do
+              var(i,j,5) = 0
+            end do
+          end do
+        end subroutine my_sub''')
+    loop = psyir.walk(Loop)[0]
+    trans = ParaTrans()
+    test_loop = psyir.copy().walk(Loop, stop_type=Loop)[0]
+    trans.apply(test_loop, {"collapse": True, "verbose": True, "force": True})
+    print(fortran_writer(test_loop.parent.parent))
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:252]: Please rm.

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py [hunk 29]
```diff
@@ -82,11 +82,187 @@ def test_paralooptrans_validate_force(fortran_reader):
...
+    do k = 1, j, 1
+      var(i,j,k) = var(i,map(j),k)
+    enddo
+  enddo
+enddo
+''' in fortran_writer(test_loop.parent.parent)
+
+    # Also it won't collapse if the loop inside is not perfectly nested,
+    # regardless of the force option, but we consider this expected and
+    # verbose won't log the reason.
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:232]: We might consider it expected but a user might not, especially if the loop body is large and hard to follow. I suggest we therefore log the reason for this too.

File: src/psyclone/tests/psyir/transformations/transformations_test.py [hunk 30]
```diff
@@ -355,77 +396,6 @@ def test_omplooptrans_properties():
...
-        do jk = 2, jpkm1, 1
-          do jj = 2, jpjm1, 1
-            do ji = 2, jpim1, 1
-              total = total + zwt(ji,jj,jk)
-            enddo
-          enddo
-        enddo''')
-    with pytest.raises(TransformationError) as err:
-        omplooptrans.validate(loops[0])
-    assert ("Transformation Error: Dependency analysis failed with the "
```
Comment (arporter) [inline src/psyclone/tests/psyir/transformations/transformations_test.py:411]: Presumably, this functionality of the DA is tested as part of the DA suite?

File: src/psyclone/transformations.py [hunk 31]
```diff
@@ -1612,13 +1613,21 @@ def validate(self, node_list, options=None):
...
                 raise TransformationError(
                     f"The provided 'default_present' option must be a "
                     f"boolean, but found '{options['default_present']}'."
                 )
+        for node in node_list:
+            for call in node.walk(Call):
+                if not call.is_available_on_device():
+                    raise TransformationError(
+                        f"'{call.routine.name}' is not available on the "
+                        f"accelerator device, and therefore it can not "
```
Comment (arporter) [inline src/psyclone/transformations.py:1629]: s/it can not ne enclosed in a/it cannot be called within an/

**Review by arporter — CHANGES_REQUESTED** [review]
Good stuff. It's really nice to see some functionality heading 'upstream' from the examples directory :-)

Pretty much all minor things. I'll check the integration tests next time.
No. of warnings for the Ref. Guide is up to 20 but they all seem to be for the usual reasons.

File: examples/nemo/scripts/omp_gpu_trans.py [hunk 32]
```diff
@@ -89,13 +89,16 @@ def trans(psyir):
 
         enhance_tree_information(subroutine)
 
+        # TODO #2082: bug in hoisting statements in loops with CodeBlocks
+        enable_hoisting = False if subroutine.name == "sto_par_init" else True
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/nemo/scripts/omp_gpu_trans.py:93]: Ups :)

File: examples/nemo/scripts/utils.py (hunk 1 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py:271]: Done

File: examples/nemo/scripts/utils.py (hunk 2 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py:272]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 5 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:94]: Removed them (and fixed typo)

I am unsure what you mean in the second comment. It collapses 'x' perfectly nested loops, it has no direction.

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 6 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:101]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 7 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:106]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 8 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:107]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 9 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:137]: Yep

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 10 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:149]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 11 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:151]: Ups, fixed :)

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 12 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:169]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 13 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:210]: Done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 15 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:250]: Yes, done

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 17 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py:266]: Done

File: src/psyclone/psyir/transformations/omp_target_trans.py (hunk 20 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/omp_target_trans.py:94]: Ups, copy-pasting mistake. Fixed now.

File: src/psyclone/psyir/transformations/omp_target_trans.py (hunk 21 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/omp_target_trans.py:109]: Done

File: src/psyclone/psyir/transformations/omp_target_trans.py (hunk 22 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/omp_target_trans.py:110]: Done

File: src/psyclone/psyir/transformations/omp_target_trans.py (hunk 23 above)
Comment (sergisiso) [reply src/psyclone/psyir/transformations/omp_target_trans.py:101]: Done

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py (hunk 24 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:107]: Done

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py (hunk 25 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:167]: Fixed

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py (hunk 26 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:168]: Done

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py (hunk 28 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:252]: Removed

File: src/psyclone/tests/psyir/transformations/transformations_test.py (hunk 30 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/transformations_test.py:411]: It is, it's one of the test cases in "test_scalar_parallelise"

File: src/psyclone/transformations.py (hunk 31 above)
Comment (sergisiso) [reply src/psyclone/transformations.py:1629]: Done

File: src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py (hunk 29 above)
Comment (sergisiso) [reply src/psyclone/tests/psyir/transformations/parallel_loop_trans_test.py:232]: Ok done

File: examples/nemo/scripts/utils.py (hunk 4 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py]: Removing this did not decrease the performance*.

* It could have been done for a large number of OpenMP threads in the CPU for the ECMWF case. But I prefer the better reliability now that we target more versions and models.

I also created #2671 for it, but @arporter points out there is other solutions.

File: examples/nemo/scripts/utils.py (hunk 3 above)
Comment (sergisiso) [reply examples/nemo/scripts/utils.py]: Commenting this out did not decrease the performance*, so I removed them now.

* It could have been done for a large number of OpenMP threads in the CPU for the ECMWF case. But I prefer the better reliability now that we target more versions and models.

File: examples/nemo/scripts/omp_gpu_trans.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/nemo/scripts/omp_gpu_trans.py]: Completely gone now since the fix has been merged first.

**Comment by sergisiso** [comment]
@arporter This is ready for another look

File: examples/nemo/scripts/omp_gpu_trans.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (arporter) [reply examples/nemo/scripts/omp_gpu_trans.py]: Oh yes, even better :-)

File: src/psyclone/psyir/transformations/parallel_loop_trans.py (hunk 5 above)
Comment (arporter) [reply src/psyclone/psyir/transformations/parallel_loop_trans.py]: Yes, don't know what I was getting confused about there.

**Review by arporter — APPROVED** [review]
All requested changes have been made, thanks.
Ref. guide is back down to 17 warnings now.
I've discovered that OMPTaskLoopTrans contains a TODO for #1370. To fix it will require extending the OMPTaskloopDirective node to support collapse so that's not something for this PR. However, it does mean we can't close #1370 so I'll amend the issue description.
I'll also re-launch the integration tests, just to be on the safe side.

**Comment by arporter** [comment]
Integration was green.
